lib/fs.js

```diff
@@ -1404,34 +1406,59 @@ function mkdirSync(path, options) {
   }
 }
 
-// TODO(Ethan-Arrowood): Make this iterative too
-function readdirSyncRecursive(path, origPath, options) {
-  nullCheck(path, 'path', true);
-  const ctx = { path };
-  const result = binding.readdir(pathModule.toNamespacedPath(path),
-                                 options.encoding, !!options.withFileTypes, undefined, ctx);
-  handleErrorFromBinding(ctx);
-  return options.withFileTypes ?
-    getDirents(path, result).flatMap((dirent) => {
-      return [
-        dirent,
-        ...(dirent.isDirectory() ?
-          readdirSyncRecursive(
-            pathModule.join(path, dirent.name),
-            origPath,
-            options,
-          ) : []),
-      ];
-    }) :
-    result.flatMap((ent) => {
-      const innerPath = pathModule.join(path, ent);
-      const relativePath = pathModule.relative(origPath, innerPath);
-      const stat = binding.internalModuleStat(innerPath);
-      return [
-        relativePath,
-        ...(stat === 1 ? readdirSyncRecursive(innerPath, origPath, options) : []),
-      ];
-    });
+/**
+ * An iterative algorithm for reading the entire contents of the `basePath` directory.
+ * This function does not validate `basePath` as a directory. It is passed directly to
+ * `binding.readdir` after a `nullCheck`.
+ * @param {string} basePath
+ * @param {{ encoding: string, withFileTypes: boolean }} options
+ * @returns {string[] | Dirent[]}
+ */
+function readdirSyncRecursive(basePath, options) {
+  nullCheck(basePath, 'path', true);
+
+  const readdirResults = [];
+  const pathsQueue = [basePath];
+
+  function read(path) {
+    const ctx = { path };
+    const readdirResult = binding.readdir(
+      pathModule.toNamespacedPath(path),
+      options.encoding,
+      Boolean(options.withFileTypes),
+      undefined,
+      ctx,
+    );
```

Comment:
Just noting some potential optimizations if you're interested in making them in this PR:

* We should be able to create `ctx` in the parent scope and reuse the object (`ctx.path = path;`) to avoid creating a new object for every call to `read()`.
* `Boolean(options.withFileTypes)` could be done once in the parent scope.
* `options.encoding` could be looked up and saved in a parent scope variable as well.